### PR TITLE
Add support for filters to T-Test aggregation

### LINKED ISSUE #53692: Add Student's t-test aggregation support
I would like to discuss adding a multivalued metrics aggregation that will apply unpaired and paired two-sample t-tests to two samples selected based on filters or fields or a combination of both. 

So, unpaired t-test might look like this:
```
GET logs/_search
{
  "size": 0,
  "aggs" : {
    "test" : {
      "t_test" : {
        "filters" : [
          { "match" : { "group" : "A" }},
          { "match" : { "group" : "B" }}
        ],
        "field": "value"
      }
    }
  }
}
```

The paired t-test might look something like this:

```
GET logs/_search
{
  "size": 0,
  "aggs" : {
    "test" : {
      "t_test" : {
        "fields" : ["before", "after"]
      }
    }
  }
}
```

We can also add support for scripts. 

The type of the test can be specified by the user with defaults based on the presence of absence of filters. We can support a `type` parameter that can be specified as `paired` (default and only supported if filters are not present), `homoscedastic` (equal variance) or `heteroscedastic` (unequal variance, default if filters are present.

The output will be a typical metrics aggregation with t and p values.

Alternatively, we can implement this as a pipeline aggregation, but in this case it will simplify implementation, but might make usage a bit more difficult and can complicate kibana adoption. We can also consider implementing it as both pipeline and metric aggregation similar to stats. 

cc: @jtibshirani, @polyfractal

### PR DESCRIPTION
**Pull request by imotov**
Adds support for filters to T-Test aggregation. The filters can be used to
select populations based on some criteria and use values from the same or
different fields.

Closes #53692

### CODE REVIEW

**Comment by imotov** [comment]
@elasticmachine update branch

File: x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/ttest/TTestAggregatorFactory.java [hunk 1]
```diff
@@ -63,13 +76,46 @@ protected Aggregator doCreateInternal(SearchContext searchContext,
         }
         switch (testType) {
             case PAIRED:
+                if (filterA != null || filterB != null) {
+                    throw new IllegalArgumentException("Paired t-test doesn't support filters");
+                }
                 return new PairedTTestAggregator(name, numericMultiVS, tails, format, searchContext, parent, metadata);
             case HOMOSCEDASTIC:
-                return new UnpairedTTestAggregator(name, numericMultiVS, tails, true, format, searchContext, parent, metadata);
+                return new UnpairedTTestAggregator(name, numericMultiVS, tails, true, this::getWeights, format, searchContext, parent,
```
Comment (polyfractal) [inline x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/ttest/TTestAggregatorFactory.java:84]: Should we validate/throw an exception/something if people specify an unpaired type, but don't provide filters?  Not sure where to put that validation logic but it seems we should disallow folks from basically running an unpaired test on two match-all's?

**Review by polyfractal — COMMENTED** [review]
Left a question about validation, but otherwise looks good.   Unpaired fits in nicely, I like how this turned out :)

File: x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/ttest/TTestAggregatorFactory.java (hunk 1 above)
Comment (imotov) [reply x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/ttest/TTestAggregatorFactory.java:84]: I was thinking about support for something like this: 
```
{"before": [1,2,3], "after":[6,7,8,9,10]}
```
 It currently works for unpaired, fails on paired and doesn't require filters.

**Review by polyfractal — APPROVED** [review]
LGTM pending what we discussed offline (failing if field is same between A/B and there is no filter) :+1:

**Comment by imotov** [comment]
@elasticmachine update branch

**Pull request merged**